bootstrap.go

```diff
@@ -147,38 +154,107 @@ func checkBootstrapped(s Session) (bool, error) {
 	return v, nil
 }
 
-func checkBootstrappedVar(s Session) (bool, error) {
+// Get variable value from mysql.tidb table.
+// Those variables are used by TiDB server.
+func getTiDBVar(s Session, name string) (types.Datum, error) {
 	sql := fmt.Sprintf(`SELECT VARIABLE_VALUE FROM %s.%s WHERE VARIABLE_NAME="%s"`,
-		mysql.SystemDB, mysql.TiDBTable, bootstrappedVar)
+		mysql.SystemDB, mysql.TiDBTable, name)
 	rs, err := s.Execute(sql)
 	if err != nil {
-		if infoschema.ErrTableNotExists.Equal(err) {
-			return false, nil
-		}
-		return false, errors.Trace(err)
+		return types.Datum{}, errors.Trace(err)
 	}
-
 	if len(rs) != 1 {
-		return false, errors.New("Wrong number of Recordset")
+		return types.Datum{}, errors.New("Wrong number of Recordset")
 	}
 	r := rs[0]
 	row, err := r.Next()
 	if err != nil || row == nil {
-		return false, errors.Trace(err)
+		return types.Datum{}, errors.Trace(err)
 	}
+	return row.Data[0], nil
+}
 
-	isBootstrapped := row.Data[0].GetString() == bootstrappedVarTrue
+func checkBootstrappedVar(s Session) (bool, error) {
+	d, err := getTiDBVar(s, bootstrappedVar)
+	if err != nil {
+		if infoschema.ErrTableNotExists.Equal(err) {
+			return false, nil
+		}
+		return false, errors.Trace(err)
+	}
+	isBootstrapped := d.GetString() == bootstrappedVarTrue
 	if isBootstrapped {
 		// Make sure that doesn't affect the following operations.
-
 		if err = s.CommitTxn(); err != nil {
 			return false, errors.Trace(err)
 		}
 	}
-
 	return isBootstrapped, nil
 }
 
+// When the system is boostrapped by low version TiDB server, we should do some upgrade works.
+// For example, add new system variables into mysql.global_variables table.
+func upgrade(s Session) error {
+	ver, err := getBootstrapVersion(s)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	if ver >= currentBootstrapVersion {
+		// It is already bootstrapped/upgraded by a higher version TiDB server.
+		if err1 := s.CommitTxn(); err1 != nil {
+			// Make sure that doesn't affect the following operations.
+			return errors.Trace(err1)
+		}
+		return nil
+	}
+	// Do upgrade works.
+	var sql string
+	if ver < 2 {
```

Comment:
Extracted the actual upgrading work as method would be more cleaner.
Like `upgradeToVer2`.